File: andre/examples/go/module6/sqrt.go
package main

import "fmt"

func sqrt(x float64) float64 {
	average := func(x float64, y float64) float64 {
		return (x + y) / 2
	}
	abs := func(x float64) float64 {
		if x < 0 {
			return -x

		} else {
			return x
		}
	}
	square := func(x float64) float64 {
		return x * x
	}
	goodEnough := func(guess float64) bool {
		return abs(square(guess)-x) < 0.001
	}
	improve := func(guess float64) float64 {
		return average(guess, (x / guess))
	}
	var sqrtIter func(float642 float64) float64 //non-orthogonal function recursion
	sqrtIter = func(guess float64) float64 {
		if goodEnough(guess) {
			return guess
		} else {
			return sqrtIter(improve(guess))
		}
	}
	return sqrtIter(1.0)
}

func main() {
	fmt.Println(sqrt(9))
}

File: andre/examples/javascript/module3/sqrt.js

let sqrt = function (x) {
    let avg = function (x, y) {
        return (x + y) / 2;
    }
    let abs = function (x) {
        return x < 0? -x : x;
    }
    let square = function (x) {
        return x * x;
    }
    let good_enough = function (guess) {
        return abs(square(guess) - x) < 0.001;
    }
    let improve = function(guess) {
        return avg(guess, (x / guess));
    }
    let sqrt_iter = function (guess) {
        if (good_enough(guess)) {
            return guess;
        } else {
            return sqrt_iter(improve(guess));
        }
    }
    return sqrt_iter(1.0);
}